### PR TITLE
push filter down traverse rule

### PR DESCRIPTION
**Pull request by jievince**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: tests/tck/features/match/Base.feature [hunk 1]
```diff
@@ -495,15 +495,15 @@ Feature: Basic match
...
+    # When executing query:
+    # """
+    # match (:player{name:"Tony Parker"})-[r]->() where exists(r.likeness) return r, exists({a:12}.a)
+    # """
+    # Then the result should be, in any order, with relax comparison:
+    # | r                                                            | exists({a:12}.a) |
+    # | [:like "Tony Parker"->"LaMarcus Aldridge" @0 {likeness: 90}] | true             |
+    # | [:like "Tony Parker"->"Manu Ginobili" @0 {likeness: 95}]     | true             |
+    # | [:like "Tony Parker"->"Tim Duncan" @0 {likeness: 95}]        | true             |
     When executing query:
```
Comment (jievince) [inline tests/tck/features/match/Base.feature:507]: Affected by https://github.com/vesoft-inc/nebula/pull/4815

File: tests/tck/features/match/Base.IntVid.feature [hunk 2]
```diff
@@ -391,15 +391,15 @@ Feature: Basic match
...
+    # When executing query:
+    # """
+    # MATCH (:player{name:"Tony Parker"})-[r]->() where exists(r.likeness) return r, exists({a:12}.a)
+    # """
+    # Then the result should be, in any order, with relax comparison:
+    # | r                                                            | exists({a:12}.a) |
+    # | [:like "Tony Parker"->"LaMarcus Aldridge" @0 {likeness: 90}] | true             |
+    # | [:like "Tony Parker"->"Manu Ginobili" @0 {likeness: 95}]     | true             |
+    # | [:like "Tony Parker"->"Tim Duncan" @0 {likeness: 95}]        | true             |
     When executing query:
```
Comment (jievince) [inline tests/tck/features/match/Base.IntVid.feature:403]: Affected by https://github.com/vesoft-inc/nebula/pull/4815

File: src/graph/optimizer/rule/PushFilterDownProjectRule.cpp [hunk 3]
```diff
@@ -24,6 +24,19 @@ PushFilterDownProjectRule::PushFilterDownProjectRule() {
   RuleSet::QueryRules().addRule(this);
 }
 
+bool PushFilterDownProjectRule::checkColumnExprKind(const Expression* expr) {
```
Comment (jievince) [inline src/graph/optimizer/rule/PushFilterDownProjectRule.cpp:27]: Copied from https://github.com/vesoft-inc/nebula/pull/4956

**Comment by Shylock-Hg** [comment]
`ent` version has support it.

**Review by yixinglu — COMMENTED** [review]
excellect!

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  if (!OptRule::match(ctx, matched)) {
+    return false;
+  }
+  DCHECK_EQ(matched.dependencies[0].dependencies[0].node->node()->kind(),
+            PlanNode::Kind::kTraverse);
+  auto traverse =
+      static_cast<const Traverse*>(matched.dependencies[0].dependencies[0].node->node());
+  auto step = traverse->stepRange();
+  // step == nullptr means one step.
+  return step == nullptr;
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:51]: For one step, the query looks like:
```
match (v1)-[e1]->(v2) where e1.likeness > 78
```
For multi step, the query should look like:
```
match (v1)-[e1*1..3]->(v2) where all(i in e1 where i.likeness > 78) 
```
The latter is rarely used currently, so just leave it for future.

**Comment by jievince** [comment]
> `ent` version has support it.

Which rule? I don't find it.

**Comment by jievince** [comment]
> `ent` version has support it.

They are not the same rule.

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+                                       {Pattern::create(PlanNode::Kind::kTraverse)})});
+  return pattern;
+}
+
+bool PushFilterDownTraverseRule::match(OptContext* ctx, const MatchedResult& matched) const {
+  if (!OptRule::match(ctx, matched)) {
+    return false;
+  }
+  DCHECK_EQ(matched.dependencies[0].dependencies[0].node->node()->kind(),
+            PlanNode::Kind::kTraverse);
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:46]: it's looks not very Intuitive. So I want to keep it that way.

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+PushFilterDownTraverseRule::PushFilterDownTraverseRule() {
+  RuleSet::QueryRules().addRule(this);
+}
+
+const Pattern& PushFilterDownTraverseRule::pattern() const {
+  static Pattern pattern =
+      Pattern::create(PlanNode::Kind::kFilter,
+                      {Pattern::create(PlanNode::Kind::kAppendVertices,
+                                       {Pattern::create(PlanNode::Kind::kTraverse)})});
+  return pattern;
```
Comment (czpmango) [inline src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:38]: Maybe we can implement this by splitting it into two rules: PushFilterDownAppendVertices and PushFilterIntoTraverse.

Comment (czpmango) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:38]: Just mark it here. You can do refactoring later.

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp (hunk 4 above)
Comment (czpmango) [inline src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:51]:
```
bool Traverse::isOneStep() {
  return !step;
}
```

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+  auto* avGroupNode = matched.dependencies[0].node;
+  auto* av = static_cast<graph::AppendVertices*>(avGroupNode->node());
+  auto& avColNames = av->colNames();
+  auto& nodeAlias = avColNames.back();
+  UNUSED(nodeAlias);
+
+  auto* tvGroupNode = matched.dependencies[0].dependencies[0].node;
+  auto* tv = static_cast<graph::Traverse*>(tvGroupNode->node());
+  auto& tvColNames = tv->colNames();
+  auto& edgeAlias = tvColNames.back();
```
Comment (czpmango) [inline src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:70]: Hard code...?
Maybe this: 
```
const auto& Traverse::getEdgeAlias() const {
    return colNames.back();
}
```

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp (hunk 4 above)
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:51]: ok

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp (hunk 7 above)
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:70]: > Hard code...? Maybe this:
> 
> ```
> const auto& Traverse::getEdgeAlias() const {
>     return colNames.back();
> }
> ```

Good job

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp (hunk 6 above)
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:38]: Leave it for the future

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+    newAvGroupNode = newAvGroup->makeGroupNode(newAv);
+    newFilterGroupNode->dependsOn(newAvGroup);
+    newFilter->setInputVar(newAv->outputVar());
+  } else {
+    newAvGroupNode = OptGroupNode::create(ctx, newAv, filterGroup);
+    newAv->setOutputVar(filter->outputVar());
+  }
+
+  auto* eFilter = tv->eFilter();
+  Expression* newEFilter = nullptr;
```
Comment (czpmango) [inline src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:132]: eFilter?LogicalExpression::makeAnd(pool, newfilterPicked, eFilter->clone()):newFilterPicked;

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp (hunk 6 above)
Comment (czpmango) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp:38]: The following scenarios are not currently covered by this optimization:
```
match (v:player)-[e:like]->() where e.likeness>80  return e.likeness
```

**Review by czpmango — DISMISSED** [review]
LGTM

File: tests/tck/features/match/Base.IntVid.feature (hunk 2 above)
Comment (yixinglu) [reply tests/tck/features/match/Base.IntVid.feature:403]: This PR has merged, could these comment out tests be enabled ?

File: src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  auto& edgeAlias = tvColNames.back();
+
+  auto qctx = ctx->qctx();
+  auto pool = qctx->objPool();
+
+  // Pick the expr looks like `$-.e[0].likeness
+  auto picker = [&edgeAlias](const Expression* e) -> bool {
+    auto varProps = graph::ExpressionUtils::collectAll(e,
+                                                       {Expression::Kind::kTagProperty,
+                                                        Expression::Kind::kEdgeProperty,
```
[the remark this answers is not in the document]
Comment (jievince) [reply src/graph/optimizer/rule/PushFilterDownTraverseRule.cpp]: ok

**Review by czpmango — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Pull request merged**